### ISSUE TITLE
Removing Python 2 code

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#17943]
Now that we have a 1.5 branch ready, we can start doing real work to drop Python 2 support. Before we jump in, I'd like to collect a list of things that should be done.

- Remove all Python 2 related things from compatibility, and from any file that imports them from compatibility.
- Remove Python 2 testing from Travis
- Grep the codebase for mentions of Python 2 compatibility things and remove them
- Fix issues mentioned in the ["Dropping Python 2"](https://github.com/sympy/sympy/labels/Dropping%20Python%202) label. 
- Add something to `sympy/__init__.py` alerting users of Python 2 that it is no longer supported. That fill will need to stay Python 2 compatible (which won't be hard). 
- Make sure setup.py is set up correctly so that it won't install in Python 2. I forget the exact way, but there is some metadata you can put there that will make it install 1.5 when in Python 2, even after we release 1.6. 

I don't know how big these changes will be, but we should try to split things into smaller pull requests as much as possible, to avoid having a big PR that gets difficult to merge because of merge conflicts. 

I think the first thing will be to remove the testing from Travis. Once that is done, we can remove the compatibility bits one-by-one. By their nature, we don't have to remove them wholesale. Technically we don't have to remove them at all, but it will clean up a lot of cruft so obviously we need to do it.

Regarding Python 3-only things, we should open new issues for them.

**Comment by oscarbenjamin** [comment]
> I think the first thing will be to remove the testing from Travis.

This is the main thing. Once we do that I predict that within a few weeks it won't be possible to import sympy master on Python 2.7. Obviously the flipside is that it saves a bunch of work making things 2/3 compatible and responding to Py2-only test failures in each PR.

We don't need to rush into excising Py2-features from the codebase though. Potentially we might need a 1.5.1 release to fix some bugs and gratuitous changes in master might make it difficult to keep the 1.5 branch in sync.

**Comment by asmeurer** [comment]
OK. Should we put the whole thing on hold then? Or do you think we should remove Python 2 from Travis now, and wait until it is clear that 1.5 won't need any further updates before we do further changes?

**Comment by asmeurer** [comment]
Another thing we should do is keep the top-level `__init__.py` Python 2 valid, so that we can add a special message to the import that says it doesn't support Python 2. That will be more user-friendly than a random syntax error in the first file it reads, which people might assume is a bug.

**Comment by oscarbenjamin** [comment]
I think that we can wait until we're sure that we're done with any bugfixes of SymPy 1.5 which will be the last release of SymPy to support Python 2.7. Once we're done with that we drop Python 2.7 and then there will be a bunch of tasks to do with excising things from the codebase. There still isn't any rush with those - everything already works on Python 3!

**Comment by oscarbenjamin** [comment]
> Another thing we should do is keep the top-level `__init__.py` Python 2 valid, so that we can add a special message to the import that says it doesn't support Python 2.

If we set the version metadata correctly we can make it so that pip refuses to install SymPy > 1.5 on Python 2.7. That way users get the latest version that supports their setup.

**Comment by asmeurer** [comment]
Oh yeah we need to do that as well. We should do both things.

**Comment by asmeurer** [comment]
It's fine to continue having it there in Travis for now. Since it is contingent on 1.5 stuff, I'll leave it up to you for when it should be done. The main thing is we can't do any of the Python 3-only stuff as long as it is there, but none of those things are urgent (I will open new issues now to discuss them).

**Comment by asmeurer** [comment]
I pushed up a `rm-python2` branch to my fork that we can use once we are ready. Or you can redo it yourself. It's not a complicated change. I just happened to already make it before I saw your comment, so I went ahead and committed and pushed it.

**Comment by hugovk** [comment]
> * Make sure setup.py is set up correctly so that it won't install in Python 2. I forget the exact way, but there is some metadata you can put there that will make it install 1.5 when in Python 2, even after we release 1.6.

Please see PRs https://github.com/sympy/sympy/pull/17966 and https://github.com/sympy/sympy/pull/17968 to add `python_requires` to the `1.5` and `master` branches.

**Comment by asmeurer** [comment]
Python 2 support has officially been dropped in https://github.com/sympy/sympy/pull/18243, which removes the Travis builds and adds an exception if SymPy is imported in Python 2. So now we can go through and do the things I mentioned above. Clearing out a majority of the compatibility file will be a good first step. It shouldn't be difficult, although it will require making a lot of changes (`git grep 'compatibility import` shows 486 imports throughout the codebase, and almost all can be removed now). There are a couple of things in that file that are there for newer versions of Python 3 (we support 3.5-3.8), and those should be kept. 

We can also independently start to work on the other "dropping Python 2" issues, like adding new Python 3 syntax like keyword-only arguments to places.

**Comment by namannimmo10** [comment]
After #18264, I'm willing to work on this one:
> (`git grep 'compatibility import` shows 486 imports throughout the codebase, and almost all can be removed now). 

Should I add [WIP] because it will take a lot of time to do that task?

**Comment by namannimmo10** [comment]
> 486 imports throughout the codebase, and almost all can be removed now

@oscarbenjamin Can't we automate this?

**Comment by oscarbenjamin** [comment]
Yes you can automate it. I'd use grep/sed. Sometimes it's easier just to write a short Python script though

**Comment by namannimmo10** [comment]
I haven't done any automation like that before.. just need some help to proceed

**Comment by asmeurer** [comment]
It can also help to use a keyboard macro if your text editor supports them. I recommend doing one compatibility import at a time (one commit per each one). A PR with even partial changes could be merged at any time.

**Comment by namannimmo10** [comment]
Thanks @asmeurer, I'm on it.

**Comment by asmeurer** [comment]
We also need to update the development workflow pages on the wiki.

**Comment by nschloe** [comment]
FWIW, I've found [pyupgrade](https://github.com/asottile/pyupgrade) to be a really handy tool.
```
find -name "*.py" | xargs pyupgrade --py3-plus
```

**Comment by smitgajjar** [comment]
Please review PR #18825, Suggest whatever is necessary!

**Comment by eagleoflqj** [comment]
Could close?
